### PR TITLE
Propagate to plane / Kalman on plane / Matriplex with support for scalar operations and VDT

### PR DESCRIPTION
**Pull request by osschar**
#### PR description:

Preliminary PR to discuss physics and computational performance of
- propagateToPlane / kalmanUpdateOnPlane
- Modifications to Matriplex: support scalar operations, usage of VDT.
- Use Matriplex modifications to re-express vectorized code for propToPlane (mostly in PropagationMPlex.icc).

#### Computational performance notes

Before Matriplex / VDT changes the track finding ran about 4-times slower. Matriplex and VDT changes brought this down by about 20% -- so we still have an enormous slowdown which is not understood at all.

Further, the Matriplex vectorization of components by use of Matriplex scalars needs to be re-checked for the new operators and for VDT. Matevz examined the assembler code for the MiniPropagators which used the prototype implementation of this concept (also including some VDT functions ... but not embedded in Matriplex).

##### Things to check / look at 

1. Compare vtune/igProf profiles for propToPlane vs. the standard implementation. Figure out where time is actually spent. Giuseppe is still working on the Kalman update on plane / in local coordinates.

2. Matriplex vectorization should be reviewed / consistently tested, along with VDT (there is special pre-processor define that causes fallback from vdt::fast_xyzz() to std::xyzz() functions.
   - Not quite sure how to go about this on an "industrial scale" -- for all functions / operators and for all usages in specific CMSSW functions.
   - I have not added simd pragmas ... they were not necessary in the MiniPropagators test. But maybe they can help by triggering some vectorization reports in case they fail.

4. Known problems / issues
   - Mario reported a CMSSW problem with static_assert assuring float/double Mplex type for VDT - I only ran standalone builds with gcc-13 on Fedora 39.
   - unary minus operator is giving me some trouble (member function without an argument didn't work, had to define it outside of the class) -- maybe I did something silly.

### CODE REVIEW

File: RecoTracker/MkFitCore/src/KalmanUtilsMPlex.cc [hunk 1]
```diff
@@ -1103,23 +1240,598 @@ namespace mkfit {
...
+    MPlexHV u;
+    RotateVectorOnPlane(rot, un, u);
+    MPlexHV v;
+    RotateVectorOnPlane(rot, vn, v);
+    MPlex55 jacCurv2Loc(0.f);
+    for (int n = 0; n < N_proc; ++n) {//fixme vectorize
+      // fixme? //(pf.use_param_b_field ? 0.01f * Const::sol * Config::bFieldFromZR(psPar(n, 2, 0), hipo(psPar(n, 0, 0), psPar(n, 1, 0))) : 0.01f * Const::sol * Config::Bfield);
+      const float bF = 0.01f * Const::sol * Config::Bfield;
+      const float qh2 = bF * lp(n,0,0);
+      const float t1r = std::sqrt(1. + lp(n,0,1)*lp(n,0,1) + lp(n,0,2)*lp(n,0,2))*pzSign(n,0,0);
```
Comment (slava77) [inline RecoTracker/MkFitCore/src/KalmanUtilsMPlex.cc:1429]:
```suggestion
      const float t1r = std::sqrt(1.f + lp(n,0,1)*lp(n,0,1) + lp(n,0,2)*lp(n,0,2))*pzSign(n,0,0);
```
?

File: RecoTracker/MkFitCore/src/KalmanUtilsMPlex.cc [hunk 2]
```diff
@@ -1009,15 +1137,17 @@ namespace mkfit {
 
   void kalmanUpdatePlane(const MPlexLS& psErr,
                          const MPlexLV& psPar,
+			                   const MPlexQI& Chg,
```
Comment (slava77) [inline RecoTracker/MkFitCore/src/KalmanUtilsMPlex.cc:1140]:
```suggestion
			                   const MPlexQI& chg,
```
naming style

File: RecoTracker/MkFitCore/src/KalmanUtilsMPlex.cc [hunk 3]
```diff
@@ -1103,23 +1240,598 @@ namespace mkfit {
...
+	un(n,0,2) = 0.f;
+	vn(n,0,0) = -tn(n,0,2)*un(n,0,1);
+	vn(n,0,1) =  tn(n,0,2)*un(n,0,0);
+      }
+      MPlex55 jacLoc2Curv(0.f);
+      for (int n = 0; n < N_proc; ++n) {//fixme vectorize
+	// fixme? //(pf.use_param_b_field ? 0.01f * Const::sol * Config::bFieldFromZR(psPar(n, 2, 0), hipo(psPar(n, 0, 0), psPar(n, 1, 0))) : 0.01f * Const::sol * Config::Bfield);
+	const float bF = 0.01f * Const::sol * Config::Bfield;//fixme: cache?
+	const float qh2 = bF * lp_upd(n,0,0);
+	const float cosl1 = 1./vn(n,0,2);
```
Comment (slava77) [inline RecoTracker/MkFitCore/src/KalmanUtilsMPlex.cc:1660]:
```suggestion
	const float cosl1 = 1.f/vn(n,0,2);
```

File: RecoTracker/MkFitCore/src/KalmanUtilsMPlex.cc [hunk 4]
```diff
@@ -1103,23 +1240,598 @@ namespace mkfit {
...
+	const float uj = un(n,0,0)*rot(n,0,0) + un(n,0,1)*rot(n,0,1);
+	const float uk = un(n,0,0)*rot(n,1,0) + un(n,0,1)*rot(n,1,1);
+	const float vj = vn(n,0,0)*rot(n,0,0) + vn(n,0,1)*rot(n,0,1) + vn(n,0,2)*rot(n,0,2);
+	const float vk = vn(n,0,0)*rot(n,1,0) + vn(n,0,1)*rot(n,1,1) + vn(n,0,2)*rot(n,1,2);
+	const float cosz = vn(n,0,2)*qh2;
+	jacLoc2Curv(n,0,0) = 1.;
```
Comment (slava77) [inline RecoTracker/MkFitCore/src/KalmanUtilsMPlex.cc:1666]:
```suggestion
	jacLoc2Curv(n,0,0) = 1.f;
```
although assignments should be OK

File: RecoTracker/MkFitCore/src/PropagationMPlex.icc [hunk 5]
```diff
@@ -235,48 +259,66 @@ inline void parsAndErrPropFromPathL_impl(const Tf& __restrict__ inPar,
...
-    jacCCS2Curv(n, 2, 4) = 1.f;
-    jacCCS2Curv(n, 3, 0) = -sinPin[n - nmin];
-    jacCCS2Curv(n, 3, 1) = cosPin[n - nmin];
-    jacCCS2Curv(n, 4, 0) = -cosPin[n - nmin] * cosT[n - nmin];
-    jacCCS2Curv(n, 4, 1) = -sinPin[n - nmin] * cosT[n - nmin];
-    jacCCS2Curv(n, 4, 2) = sinT[n - nmin];
-  }
+  MPlex56 jacCCS2Curv(0.0f);
+  jacCCS2Curv.aij(0, 3) = mpt::negate_if_ltz(sinT, inChg);
+  jacCCS2Curv.aij(0, 5) = mpt::negate_if_ltz(cosT, inChg);
```
Comment (slava77) [inline RecoTracker/MkFitCore/src/PropagationMPlex.icc:264]: source in `TrackState::jacobianCCSToCurvilinear` has `    jac(0, 5) = charge * cosT * invpt;`

Comment (slava77) [reply RecoTracker/MkFitCore/src/PropagationMPlex.icc:264]: > source in `TrackState::jacobianCCSToCurvilinear` has ` jac(0, 5) = charge * cosT * invpt;`

@cerati which one do you think is a bug?

Comment (cerati) [reply RecoTracker/MkFitCore/src/PropagationMPlex.icc:264]: I am not familiar with `mpt::negate_if_ltz`... maybe @osschar ?

Comment (cerati) [reply RecoTracker/MkFitCore/src/PropagationMPlex.icc:264]: I think my version before this change was the same as `TrackState::jacobianCCSToCurvilinear`

File: RecoTracker/MkFitCore/src/KalmanUtilsMPlex.cc (hunk 4 above)
Comment (dan131riley) [reply RecoTracker/MkFitCore/src/KalmanUtilsMPlex.cc]: https://github.com/trackreco/cmssw/blob/6f9a138a316977d892e404eb3e22b627e07ea1ce/RecoTracker/MkFitCore/src/Matriplex/Matriplex.h#L78-L85

Brute forcing the matrix to positive...

File: RecoTracker/MkFitCore/src/PropagationMPlex.icc (hunk 5 above)
Comment (slava77) [reply RecoTracker/MkFitCore/src/PropagationMPlex.icc:264]: my point is that the `1/pT` is missing

Comment (cerati) [reply RecoTracker/MkFitCore/src/PropagationMPlex.icc:264]: it was not missing in my version (you can see it if you look at the code that was removed)

Comment (osschar) [reply RecoTracker/MkFitCore/src/PropagationMPlex.icc:264]: oh rats, it's my fault again :)
or maybe it's better because of it ;)

**Comment by osschar** [comment]
Changes added to #151, this is now obsolete, closing.

**Pull request closed without merging**